fix comparison count for format_version=3 indexes

ajkr opened the pull request:
Summary: In index blocks since `format_version=3`, user keys are written
rather than internal keys. When reading such blocks, the comparator is
obtained via `InternalKeyComparator::user_comparator()`. We need to
wrap that function's result in a `UserComparatorWrapper` in order for
its comparisons to be counted in `PerfContext::user_key_comparison_count`.

Test Plan: ran db_bench and verified
`PerfContext::user_key_comparison_count` became larger.

siying commented on db/dbformat.h:
```diff
@@ -214,7 +214,7 @@ class InternalKeyComparator
 
   const Comparator* user_comparator() const {
-    return user_comparator_.user_comparator();
+    return &user_comparator_;
```
There seems to be a lot of caller to this functions. Are we fixing accounts for all of them, or is there a chance we overcount somewhere?

siying replied:
I think about it again. Can we create a new function that returns the wrapper instead? In that case, the user of the function will be explicit about whether the comparisons are counted underlying or not. And it can also potentially avoid one virtual function call.

ajkr replied:
Do you mind helping me understand what kind of comparison should not be counted when `PerfContext` is enabled? I am not that familiar with this wrapper so am still naively hoping we can land in one of the following simple situations:

1. the wrapper is cheap and we allow it to count all comparisons
2. the wrapper is expensive so we get rid of it entirely

ajkr replied:
Never mind, I saw there's an existing pattern to follow (call `UserComparatorWrapper(rep_->internal_comparator.user_comparator()).Compare()`). Will follow that pattern.

ajkr commented:
@siying addressed your comment, ready for another look.

By the way, I am thinking we should get rid of the `UserComparatorWrapper` entirely. It is too confusing to have a subclass of Comparator while not wanting to make virtual function calls on it. Not to mention it's hard to know where to wrap it or whether it's already been wrapped.

An alternative is just use macros. There would be a `COMPARE()` macro that handles bumping the counter followed by invoking the comparison. What do you think? (This would be a future PR.)
